0013-use-backfill-to-let-mv-on-mv-stream-again.md

```diff
@@ -0,0 +1,51 @@
+---
+feature: use_backfill_to_let_mv_on_mv_stream_again
+authors:
+  - "DylanChen"
+start_date: "2022/11/04"
+---
+
+# Use Backfill To Let Mv On Mv Stream Again
+
+## Summary
+
+Currently, if we want to create a mv A on another mv B, we need to scan mv B with a snapshot at first which will cause we need to buffer all stream upadte message from mv B during the snapshot reading. This unbounded buffer will consume unbounded memory and also make the streaming processing stuck in some way. This RFC try to use backfill to save mv-on-mv from buffering all update steam message.
+
+## Motivation
+
+There is a paper *Online, Asynchronous Schema Change in F1* introduced by google used to solve the online schema change problem of distributed system. They use online schema change to solve the creating index problem which somehow is similar to our creating index case or even our creating materialized view case. Online schema change use a backfill operation to backfill all missing key-value pairs corresponding to the new element and this backfill operation can run concurrently with any upstream update. This idea can be used to our system. In this way, we can eliminate the unbounded buffer of upstream update.
+
+## Design
+
+We use creating index as a example to illustrate the idea, however, we can apply it to create mv too.Suppose that we have a table T which contains 10,000 rows. We want to create an index on it. This index is actually also a materialized view. The backfill operations work like as following.
+
+### Backfill
+
+- Use a variable `current_pos` to keep track the position (pk) we have backfilled which initialized as 0.
```

Comment:
Does this work on a snapshot or we can move to the next snapshot?

Reply:
I assume pk is sorted in an ascend way. If the first pk is "b" scanned by the backfill, the current_pos is "c", and there's an update from "a" (which is smaller than current_pos), what should we do?

Reply:
> I assume pk is sorted in an ascend way. If the first pk is "b" scanned by the backfill, the current_pos is "c", and there's an update from "a" (which is smaller than current_pos), what should we do?

Forward update "a" to the downstream because it is smaller than current_pos. Because if current_pos is "b", we can ensure that any row with pk smaller than "b" have already been sent to downstream.